[Build] Configure and satisfy JSHint

VWoeltjen opened the pull request:
Addresses #671 by adding JSHint rules and enforcing them as part of the `verify` task.

Summary of changes:

* Configured JSHint; preferring to identify potential errors aggressively
  * Enabled all non-deprecated boolean enforcement options
  * Did not enable any relaxed options
  * Declared `define` as a global and set `strict` to `implied` to reduce boilerplate
  * Declared `browser` environment (and `jasmine` environment for specs)
  * Did not set any `max...` options (e.g. `maxstatements`); I'd actually like to, but don't have a good sense of what the right values would be, and want to establish a more neutral baseline set of JSHint options.
* Updated existing scripts to satisfy JSHint as configured
  * Removed `"use strict"` and `/*global define*/` (automated using script in the `scripts` directory)
  * Removed unused variables
  * Normalized position of operators in multiline expressions
  * Misc. changes (`for...in` to `Object.keys(...).forEach`)

@larkin / @akhenry - this is also an opportunity to evaluate the set of [JSHint rules](http://jshint.com/docs/options/#jasmine) we use. For my part, I prefer strictness (allocates a larger subset of review responsibility to the machines), but understand that it has drawbacks (refactoring perfectly fine code to satisfy JSHint is a bummer.)

### Author Checklist

1. Changes address original issue? Y
2. Unit tests included and/or updated with changes? Y
3. Command line build passes? Y
4. Changes have been smoke-tested? Y

VWoeltjen commented:
Given the pretty extensive diff (the majority of which is boilerplate removal) it may be easier to look at individual commits. If there's something I can do to make review easier (e.g. rebase and squash similar commits?) I'd be happy to do so.

akhenry commented:
```
Misc. changes (for...in to Object.keys(...).forEach)
```
I'm sure you were cognizant of this when making this change, but these are not equivalent. If you're changing code from the former to the latter you need to be really careful of context and that the resulting code actually does what it's supposed to after you change it. A specific example that comes to mind is in the `EditableDomainObjectImpl` where it wraps function calls. `Object.keys` will not pick up functions, nor properties inherited through prototype. From looking at the diff I can see though that this specific instance is unchanged.

akhenry commented:
@larkin If you want to jointly review this one, I'd be happy to divide up the work?

VWoeltjen commented:
Realizing this isn't properly integrated into CircleCI config - reassigning to self to handle that

akhenry commented on platform/commonUI/edit/src/actions/SaveAction.js:
```diff
@@ -19,14 +19,12 @@
  * at runtime from the About dialog for additional information.
  *****************************************************************************/
-/*global define*/
 /*jslint es5: true */
```
I added this jslint rule in a couple of places. It's now redundant (because we're using jshint 2+ which [enables ES5 mode by default](https://jslinterrors.com/es5-option-is-now-set-per-default) ). This rule was added specifically because jslint did not like the reserved work 'catch' being used as a function name.

larkin commented:
Given the size of the diff, worth explaining my approach to this:
1. Ran diff stat and looked at number of changes per file.  Used some command line tools to generate `unique diffs`-- e.g. two files with the removal of `/*global define*/` and `'use strict';` can be considered the same change.  Was able to quickly review 644 files matching this criteria.
2. Looked at remaining files and reviewed by hand.  Took notes separately as github wont allow per-line comments on a pull request of this size.

Specific comments:

* CouchPersistenceProvider and ElasticPersistenceProvider: we're using a nonstandard `bind` function here.  Can we switch this statement to Function.prototype.bind?
* GenericSearchProvider: I used setTimeout over $timeout for performance reasons.  $timeout constructs a promise for every call and I wanted minimize memory thrashing.
* LocalStoragePersistenceProviderSpec: refactored to depend on $window, but still passing an unnecessary 4th argument to constructor.

Feedback on JSHint config:
1. I would like to change `unused` from `true` to `vars` so that it only complains about unused variables when they are declared in var statements and not as function arguments.  Otherwise, stub methods that are intended to be overridden (or that may be overridden) rely solely on jsdoc for showing their arguments.
2. Why did we remove .jshintrc?  This allows editors to read the jshint config from a file and apply it per-project.  Without it, editors & other dev tools will not enforce the project standards and it will make development harder.

Reassigning to @VWoeltjen, let me your thoughts on above points.  Any commits you add to address these points I will review individually for quicker re-review.

VWoeltjen commented:
@akhenry:

* Good callout on Object.keys versus for-in; I was conscious of the difference, and in the case where I replaced for-in with Object.keys (134452582c5c1c8e9c0d4dc150609b8b3fa2c8ca) I believe Object.keys is more correct (an object is being used to map type keys to booleans for selection state, without any prototypal inheritance at play)
* `grep`ped for jslint configurations and removed these in 934eb13813254294d75fc8052ece74be96beea8d, thanks for pointing these out

@larkin:

* Switched Couch/Elastic persistence providers to Function.prototype.bind in f6a9c90cef7ea0822731550886fbd63b6b8a508f
* Restored `setTimeout` usage in GenericSearchProvider in aa45e5344082ff24f34164f1d2cb61dd3847b945
* Removed obsolete argument from LocalStoragePersistenceSpec in 93512851828ff89686d729384e35b0201596185e
* Changed `unused` to `"vars"` (agree with the rationale you stated) in c1ae68b565bcc33c61a7f3ec9e96a54492c10a9f
* Moved config out to `.jshintrc` in a224711dce6b2d0e7bd4e87139ef18be04cb1891 (was trying to cleanup top-level files a bit and localize build configuration in the gulpfile, but hadn't considered code editors - much better to allow those to pick up on settings)

Reassigning to continue review

larkin commented:
# Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? N/A (Config/Build Change change)
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y

Say hello to a new world of linting!
